File: polyglot/envelope_v1/php/main.php
#!/usr/bin/env php
<?php
// B1 fourteen-language conformance: PHP.
//
// Invariant owned: effect_requires_authority_digest
//
// This is the load-bearing one for the authority model. An envelope that names
// a persistent effect must also name the authority envelope that permitted it.
// An effect record with no bound authority is unauthorised by construction,
// whatever the surrounding prose claims, and no amount of downstream checking
// can recover an authorisation that was never recorded.
//
// The check runs against the effect vector, and it also confirms the digest is
// well formed: an authority reference that is not a real 64-character hex
// digest cannot be resolved to an envelope, which makes it decoration.
//
// digest_verification: STDLIB. hash('sha256', ...) is built in.
//
// Usage: php main.php <contract.json> <vector.canonical> <expected.json>

declare(strict_types=1);

const INVARIANT = 'effect_requires_authority_digest';

function fail(string $message, int $code): never {
    fwrite(STDERR, "PHP: {$message}\n");
    exit($code);
}

$args = array_slice($argv, 1);
if (count($args) !== 3) {
    fail('usage: php main.php <contract.json> <vector.canonical> <expected.json>', 2);
}
[$contractPath, $vectorPath, $expectedPath] = $args;

$contractRaw = @file_get_contents($contractPath);
if ($contractRaw === false) { fail('cannot read contract', 3); }
$vectorBytes = @file_get_contents($vectorPath);
if ($vectorBytes === false) { fail('cannot read vector', 3); }
$expectedRaw = @file_get_contents($expectedPath);
if ($expectedRaw === false) { fail('cannot read expected', 3); }

$contract = json_decode($contractRaw, true);
$envelope = json_decode($vectorBytes, true);
$expected = json_decode($expectedRaw, true);

if (!is_array($contract) || !is_array($envelope) || !is_array($expected)) {
    fail('one of the inputs is not valid JSON', 3);
}

// The contract must actually assert this invariant. A consumer checking a
// property nobody declared is testing its own opinion, not B1's contract.
if (($contract['invariants'][INVARIANT] ?? null) !== true) {
    fail('contract does not assert ' . INVARIANT, 4);
}

// This consumer only means something against an effect record. Passing on a
// vector with no effect would be vacuous, and a vacuous pass looks like
// evidence without being any.
if (!array_key_exists('effect_identity', $envelope)) {
    fail(
        'vector carries no effect_identity, so it cannot demonstrate this invariant; '
        . 'point this consumer at the effect vector',
        5
    );
}

if (!array_key_exists('authority_envelope_digest', $envelope)) {
    fail(
        'effect ' . json_encode($envelope['effect_identity']) . ' names no '
        . 'authority_envelope_digest; an effect with no bound authority is '
        . 'unauthorised by construction',
        5
    );
}

$digest = $envelope['authority_envelope_digest'];
if (!is_string($digest) || preg_match('/^[0-9a-f]{64}$/', $digest) !== 1) {
    fail(
        'authority_envelope_digest is not 64 lowercase hex characters, so it cannot be '
        . 'resolved to an authority envelope; an unresolvable reference is decoration',
        5
    );
}

$actual = hash('sha256', $vectorBytes);
if ($actual !== ($expected['canonical_sha256'] ?? null)) {
    fail(
        'digest mismatch; declared ' . ($expected['canonical_sha256'] ?? 'nothing')
        . ' but the bytes hash to ' . $actual,
        5
    );
}

echo 'PHP:POSTCONDITION:' . INVARIANT . "\n";
exit(0);

File: polyglot/envelope_v1/ruby/main.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# B1 fourteen-language conformance: Ruby.
#
# Invariant owned: effect_requires_persistence_class
#
# The companion to PHP's authority check. An envelope that names a persistent
# effect must also carry the recovery class that effect was classified under:
# REVERSIBLE, COMPENSATABLE, IRREVERSIBLE or UNKNOWN. The point is the ordering,
# not the label. Recovery limits have to be known *before* the effect happens;
# an effect whose class is discovered afterwards was authorised without anyone
# knowing what undoing it would cost.
#
# UNKNOWN is deliberately a permitted value. "We could not establish the
# recovery characteristics" is an honest classification and must be recordable;
# what is not permitted is the field being absent, which records nothing while
# looking like nothing is wrong.
#
# digest_verification: STDLIB. Ruby's digest library provides SHA-256.
#
# Usage: ruby main.rb <contract.json> <vector.canonical> <expected.json>

require 'digest'
require 'json'

INVARIANT = 'effect_requires_persistence_class'
ALLOWED = %w[REVERSIBLE COMPENSATABLE IRREVERSIBLE UNKNOWN].freeze

def fail_with(message, code)
  warn "Ruby: #{message}"
  exit code
end

args = ARGV
fail_with('usage: ruby main.rb <contract.json> <vector.canonical> <expected.json>', 2) if args.length != 3

contract_path, vector_path, expected_path = args

begin
  contract = JSON.parse(File.read(contract_path, encoding: 'UTF-8'))
  vector_bytes = File.binread(vector_path)
  expected = JSON.parse(File.read(expected_path, encoding: 'UTF-8'))
rescue SystemCallError => e
  fail_with("cannot read an input: #{e.message}", 3)
rescue JSON::ParserError => e
  fail_with("an input is not valid JSON: #{e.message}", 3)
end

# The contract must actually assert this invariant. A consumer checking a
# property nobody declared is testing its own opinion, not B1's contract.
unless contract.dig('invariants', INVARIANT) == true
  fail_with("contract does not assert #{INVARIANT}", 4)
end

envelope = JSON.parse(vector_bytes.force_encoding('UTF-8'))

# This consumer only means something against an effect record. Passing on a
# vector with no effect would be vacuous, and a vacuous pass looks like evidence
# without being any.
unless envelope.key?('effect_identity')
  fail_with(
    'vector carries no effect_identity, so it cannot demonstrate this invariant; ' \
    'point this consumer at the effect vector',
    5
  )
end

unless envelope.key?('persistence_class')
  fail_with(
    "effect #{envelope['effect_identity'].inspect} carries no persistence_class; " \
    'recovery limits must be classified before the effect, not discovered afterwards',
    5
  )
end

klass = envelope['persistence_class']
unless ALLOWED.include?(klass)
  fail_with(
    "persistence_class #{klass.inspect} is not one of #{ALLOWED.join(', ')}; " \
    'UNKNOWN is a permitted and honest answer, an unrecognised label is not',
    5
  )
end

actual = Digest::SHA256.hexdigest(vector_bytes)
unless actual == expected['canonical_sha256']
  fail_with(
    "digest mismatch; declared #{expected['canonical_sha256']} but the bytes hash to #{actual}",
    5
  )
end

puts "RUBY:POSTCONDITION:#{INVARIANT}"
exit 0
